### PR TITLE
feat: add escrow pallet to vote lock KINT

### PR DESCRIPTION
**Pull request by gregdhill**
Signed-off-by: Gregory Hill <gregorydhill@outlook.com>

Implementation of Curve's [VotingEscrow.vy](https://github.com/curvefi/curve-dao-contracts/blob/master/contracts/VotingEscrow.vy) contract for stake-to-vote locking.

### CODE REVIEW

File: crates/escrow/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+    fn current_height() -> T::BlockNumber {
+        frame_system::Pallet::<T>::block_number()
+    }
+
+    fn round_height(height: T::BlockNumber) -> T::BlockNumber {
+        let span = T::Span::get();
+        (height / span) * span
+    }
+
+    fn checkpoint(who: &T::AccountId, old_locked: DefaultLockedBalance<T>, new_locked: DefaultLockedBalance<T>) {
```
Comment (sander2) [inline crates/escrow/src/lib.rs:293]: This function is a bit hard to follow, partly because of curve's naming standards. I would love to see this function in the spec, together with some explanation about exactly what it is supposed to achieve, and also when it is expected to be called. One thing in particular that I don't immediately understand, is `SlopeChanges`. I thought this was to account for additional amounts being locked when you already have some tokens locked, but then I don't understand how it is indexed only by the blocknumber (and not the account)

File: crates/escrow/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+    pub struct Pallet<T>(_);
+
+    // The pallet's dispatchable functions.
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::weight(<T as Config>::WeightInfo::create_lock())]
+        #[transactional]
+        pub fn create_lock(
+            origin: OriginFor<T>,
+            #[pallet::compact] amount: BalanceOf<T>,
```
Comment (nud3l) [inline crates/escrow/src/lib.rs:185]: I might read this wrong, but wouldn't that allow to lock any currency? We only want to allow to lock KINT on kintsugi and INTR on interlay.

File: crates/escrow/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,602 @@
+//! # Escrow Module
+//! Receive vote-escrowed tokens for locking the native currency.
```
Comment (nud3l) [inline crates/escrow/src/lib.rs:2]:
```suggestion
//! Receive vote-escrowed tokens for locking the native currency.
//! Follows the specification at https://spec.interlay.io/spec/escrow.html
```

Comment (nud3l) [reply crates/escrow/src/lib.rs:2]: Note not yet merged on the spec, but will be available there

File: crates/escrow/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+
+        Self {
+            bias,
+            slope,
+            ts: start_height,
+        }
+    }
+
+    fn balance_at<BlockNumberToBalance: Convert<BlockNumber, Balance>>(&self, height: BlockNumber) -> Balance {
+        let height_diff = BlockNumberToBalance::convert(height - self.ts);
```
Comment (nud3l) [inline crates/escrow/src/lib.rs:73]: Isn't `self.ts` in the future and thus `self.ts > height` -> doesn't this overflow?

File: crates/escrow/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+
+            // lock amount MUST be non-zero
+            ensure!(!locked_balance.amount.is_zero(), Error::<T>::InvalidAmount);
+
+            // lock duration MUST increase
+            ensure!(unlock_height > locked_balance.end, Error::<T>::InvalidHeight);
+
+            // height MUST NOT be greater than max
+            let max_period = T::MaxPeriod::get();
+            let end_height = now + max_period;
```
Comment (nud3l) [inline crates/escrow/src/lib.rs:261]: Please use safe math

File: crates/escrow/src/lib.rs (hunk 2 above)
Comment (sander2) [reply crates/escrow/src/lib.rs:185]: No, `BalanceOf` is essentially a u128. In contrast to the `Amount` struct that we use throughout much of the code, it does not include the currencyid. The used currency is defined by the `Currency` type in the config

File: crates/escrow/src/lib.rs (hunk 1 above)
Comment (gregdhill) [reply crates/escrow/src/lib.rs:293]: This pallet is a prototype port of the [vyper contract](https://github.com/curvefi/curve-dao-contracts/blob/master/contracts/VotingEscrow.vy). In the interest of time please review their code / spec for now and we will optimize later.

File: crates/escrow/src/lib.rs (hunk 4 above)
Comment (gregdhill) [reply crates/escrow/src/lib.rs:73]: No `self.ts` is likely always in the past, but I will adjust the implementation to use saturated math here.

**Review by sander2 — APPROVED** [review]
